File: tests/go/unit/identity/jwks_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// JWKS cache tests for fetch, TTL caching, and cache reset behavior.

package identity_test

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/garudex-labs/caracal/packages/identity/go"
)

func TestGetJWKSFetchesFromWellKnownEndpoint(t *testing.T) {
	identity.ResetJWKSCache()
	var gotPath string
	var gotZone string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		gotZone = r.URL.Query().Get("zone_id")
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"keys": []any{}})
	}))
	defer server.Close()

	_, err := identity.GetJWKS(server.URL, "zone1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if gotPath != "/.well-known/jwks.json" {
		t.Fatalf("expected /.well-known/jwks.json, got %q", gotPath)
	}
	if gotZone != "zone1" {
		t.Fatalf("expected zone_id=zone1, got %q", gotZone)
	}
}

func TestGetJWKSRequiresZone(t *testing.T) {
	identity.ResetJWKSCache()
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
	}))
	defer server.Close()

	if _, err := identity.GetJWKS(server.URL, ""); err == nil {
		t.Fatal("expected error for missing zone_id")
	}
	if calls != 0 {
		t.Fatalf("expected no fetch without zone_id, got %d", calls)
	}
}

func TestGetJWKSCachesPerZone(t *testing.T) {
	identity.ResetJWKSCache()
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"keys": []any{}})
	}))
	defer server.Close()

	if _, err := identity.GetJWKS(server.URL, "zone1"); err != nil {
		t.Fatalf("zone1 fetch failed: %v", err)
	}
	if _, err := identity.GetJWKS(server.URL, "zone2"); err != nil {
		t.Fatalf("zone2 fetch failed: %v", err)
	}

	if calls != 2 {
		t.Fatalf("expected separate fetches per zone, got %d", calls)
	}
}

func TestGetJWKSCachesResultAndSkipsRefetch(t *testing.T) {
	identity.ResetJWKSCache()
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"keys": []any{}})
	}))
	defer server.Close()

	if _, err := identity.GetJWKS(server.URL, "zone1"); err != nil {
		t.Fatalf("first fetch failed: %v", err)
	}
	if _, err := identity.GetJWKS(server.URL, "zone1"); err != nil {
		t.Fatalf("second fetch failed: %v", err)
	}

	if calls != 1 {
		t.Fatalf("expected 1 fetch, got %d", calls)
	}
}

func TestResetJWKSCacheForcesRefetch(t *testing.T) {
	identity.ResetJWKSCache()
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"keys": []any{}})
	}))
	defer server.Close()

	if _, err := identity.GetJWKS(server.URL, "zone1"); err != nil {
		t.Fatalf("first fetch failed: %v", err)
	}
	identity.ResetJWKSCache()
	if _, err := identity.GetJWKS(server.URL, "zone1"); err != nil {
		t.Fatalf("second fetch failed: %v", err)
	}

	if calls != 2 {
		t.Fatalf("expected 2 fetches after reset, got %d", calls)
	}
}

func TestGetJWKSRejectsNonOKStatus(t *testing.T) {
	identity.ResetJWKSCache()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusServiceUnavailable)
	}))
	defer server.Close()

	_, err := identity.GetJWKS(server.URL, "zone1")
	if err == nil {
		t.Fatal("expected error for 503 response")
	}
}

func TestGetJWKSRejectsInvalidJSON(t *testing.T) {
	identity.ResetJWKSCache()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte("{not valid json"))
	}))
	defer server.Close()

	_, err := identity.GetJWKS(server.URL, "zone1")
	if err == nil {
		t.Fatal("expected error for malformed JSON response")
	}
}

File: tests/typescript/unit/identity/jwks.test.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// TypeScript identity JWKS cache unit tests for fetch and cache behavior.

import { afterEach, describe, expect, it, vi } from 'vitest'
import { clearJwksCache, createJwksCache, getKeySet } from '../../../../packages/identity/ts/src/jwks.js'

const TTL_MS = 5 * 60 * 1000

describe('getKeySet', () => {
  afterEach(() => {
    clearJwksCache()
    vi.unstubAllGlobals()
    vi.useRealTimers()
  })

  it('fetches zone-scoped JWKS from the issuer well-known endpoint', async () => {
    const issuer = 'https://issuer-one.example/'
    const fetchMock = vi.fn().mockResolvedValue({ ok: true, json: async () => ({ keys: [] }) })
    vi.stubGlobal('fetch', fetchMock)

    const keySet = await getKeySet(issuer, 'zone1')

    expect(typeof keySet).toBe('function')
    expect(fetchMock).toHaveBeenCalledWith('https://issuer-one.example/.well-known/jwks.json?zone_id=zone1', {
      signal: expect.any(AbortSignal),
    })
  })

  it('reuses cached key sets for the same issuer and zone', async () => {
    const issuer = 'https://issuer-two.example'
    const fetchMock = vi.fn().mockResolvedValue({ ok: true, json: async () => ({ keys: [] }) })
    vi.stubGlobal('fetch', fetchMock)

    const first = await getKeySet(issuer, 'zone1')
    const second = await getKeySet(issuer, 'zone1')

    expect(first).toBe(second)
    expect(fetchMock).toHaveBeenCalledTimes(1)
  })

  it('rejects key set lookups without a zone', async () => {
    const fetchMock = vi.fn()
    vi.stubGlobal('fetch', fetchMock)

    await expect(getKeySet('https://issuer-nozone.example', '')).rejects.toThrow('zone_id required')
    expect(fetchMock).not.toHaveBeenCalled()
  })

  it('caches key sets per zone', async () => {
    const issuer = 'https://issuer-zones.example'
    const fetchMock = vi.fn().mockResolvedValue({ ok: true, json: async () => ({ keys: [] }) })
    vi.stubGlobal('fetch', fetchMock)

    await getKeySet(issuer, 'zone1')
    await getKeySet(issuer, 'zone2')

    expect(fetchMock).toHaveBeenCalledTimes(2)
    expect(fetchMock).toHaveBeenCalledWith('https://issuer-zones.example/.well-known/jwks.json?zone_id=zone2', {
      signal: expect.any(AbortSignal),
    })
  })

  it('rejects failed JWKS fetches', async () => {
    vi.stubGlobal('fetch', vi.fn().mockResolvedValue({ ok: false, status: 503 }))

    await expect(getKeySet('https://issuer-three.example', 'zone1')).rejects.toThrow('JWKS fetch failed: 503')
  })

  it('returns stale key set and triggers background revalidation after TTL', async () => {
    vi.useFakeTimers()
    const issuer = 'https://issuer-stale.example'
    const fetchMock = vi.fn().mockResolvedValue({ ok: true, json: async () => ({ keys: [] }) })
    vi.stubGlobal('fetch', fetchMock)

    const stale = await getKeySet(issuer, 'zone1')
    expect(fetchMock).toHaveBeenCalledTimes(1)

    vi.advanceTimersByTime(TTL_MS + 1)

    const returned = await getKeySet(issuer, 'zone1')
    expect(returned).toBe(stale)

    await vi.runAllTimersAsync()
    expect(fetchMock).toHaveBeenCalledTimes(2)
  })

  it('resets revalidating flag when background refresh fails', async () => {
    vi.useFakeTimers()
    const issuer = 'https://issuer-stale-fail.example'
    const fetchMock = vi.fn()
      .mockResolvedValueOnce({ ok: true, json: async () => ({ keys: [] }) })
      .mockResolvedValueOnce({ ok: false, status: 500 })
      .mockResolvedValueOnce({ ok: true, json: async () => ({ keys: [] }) })
    vi.stubGlobal('fetch', fetchMock)

    await getKeySet(issuer, 'zone1')

    vi.advanceTimersByTime(TTL_MS + 1)
    await getKeySet(issuer, 'zone1')
    await vi.runAllTimersAsync()

    vi.advanceTimersByTime(TTL_MS + 1)
    await getKeySet(issuer, 'zone1')
    await vi.runAllTimersAsync()

    expect(fetchMock).toHaveBeenCalledTimes(3)
  })

  it('fails closed when cached keys exceed max stale', async () => {
    const issuer = 'https://issuer-max-stale.example'
    let calls = 0
    const cache = createJwksCache({
      ttlMs: 0,
      maxStaleMs: 10,
      fetchImpl: async () => {
        calls++
        if (calls === 1) return new Response(JSON.stringify({ keys: [] }), { status: 200 })
        return new Response('unavailable', { status: 503 })
      },
    })

    await cache.getKeySet(issuer, 'zone1')
    await cache.getKeySet(issuer, 'zone1')
    await new Promise((resolve) => setTimeout(resolve, 15))
    await expect(cache.getKeySet(issuer, 'zone1')).rejects.toThrow('JWKS fetch failed: 503')
  })
})
